File: TypeScript/src/contracts/tunnel.ts
import { TunnelAccessControl } from './tunnelAccessControl';
import { TunnelEndpoint } from './tunnelEndpoint';
import { TunnelOptions } from './tunnelOptions';
import { TunnelPort } from './tunnelPort';
import { TunnelStatus } from './tunnelStatus';

/**
 * Data contract for tunnel objects managed through the tunnel service REST API.
 */
export interface Tunnel {
    /**
     * Gets or sets the ID of the cluster the tunnel was created in.
     */
    clusterId?: string;

    /**
     * Gets or sets the generated ID of the tunnel, unique within the cluster.
     */
    tunnelId?: string;

    /**
     * Gets or sets the optional short name (alias) of the tunnel.
     * The name must be globally unique within the parent domain, and must be a valid subdomain.
     */
    name?: string;

    /**
     * Gets or sets the description of the tunnel.
     */
    description?: string;

    /**
     * Gets or sets the tags of the tunnel.
     */
    tags?: string[];

    /**
     * Gets or sets the optional parent domain of the tunnel,
     * if it is not using the default parent domain.
     */
    domain?: string;

    /**
     * Gets or sets a dictionary mapping from scopes to tunnel access tokens.
     */
    accessTokens?: { [scope: string]: string };

    /**
     * Gets or sets access control settings for the tunnel.
     */
    accessControl?: TunnelAccessControl;

    /**
     * Gets or sets default options for the tunnel.
     */
    options?: TunnelOptions;

    /**
     * Gets or sets current connection status of the tunnel.
     */
    status?: TunnelStatus;

    /**
     * Gets or sets an array of endpoints where hosts are currently accepting
     * client connections to the tunnel.
     */
    endpoints?: TunnelEndpoint[];

    /**
     * Gets or sets a list of ports in the tunnel.
     */
    ports?: TunnelPort[];
}

File: swift/Sources/DevTunnelsClient/Contracts/Tunnel.swift
/// Data contract for tunnel objects managed through the tunnel service REST API.
public struct Tunnel: Codable, Equatable, Sendable {
    /// Cluster ID where the tunnel was created.
    public var clusterId: String?

    /// Generated unique tunnel ID within the cluster.
    public var tunnelId: String?

    /// Optional short name (alias). Globally unique within the parent domain.
    public var name: String?

    /// Description of the tunnel.
    public var description: String?

    /// Labels for the tunnel.
    public var labels: [String]?

    /// Optional parent domain (if not using the default).
    public var domain: String?

    /// Dictionary mapping from scopes to tunnel access tokens.
    public var accessTokens: [String: String]?

    /// Current connection status of the tunnel.
    public var status: TunnelStatus?

    /// Endpoints where hosts are currently accepting client connections.
    public var endpoints: [TunnelEndpoint]?

    /// Ports in the tunnel.
    public var ports: [TunnelPort]?

    public init(
        clusterId: String? = nil,
        tunnelId: String? = nil,
        name: String? = nil,
        description: String? = nil,
        labels: [String]? = nil,
        domain: String? = nil,
        accessTokens: [String: String]? = nil,
        status: TunnelStatus? = nil,
        endpoints: [TunnelEndpoint]? = nil,
        ports: [TunnelPort]? = nil
    ) {
        self.clusterId = clusterId
        self.tunnelId = tunnelId
        self.name = name
        self.description = description
        self.labels = labels
        self.domain = domain
        self.accessTokens = accessTokens
        self.status = status
        self.endpoints = endpoints
        self.ports = ports
    }
}
